### PR TITLE
Implemented PrizeDistributionFactory to move prize dist calc on-chain

### PR DESCRIPTION
**Pull request by asselstine**

### CODE REVIEW

**Comment by kamescg** [comment]
@asselstine Does this need a review or ready to merge?

I think by tomorrow this might be blocking, because `v4-testnet` scripts will need this ABI published to NPM before moving forward with the updated deploy script that will use this new contract and the Timelock contract I am working on today.

**Comment by kamescg** [comment]
Tracking temporary reference to `interface IPrizeDistributionFactory` in https://github.com/pooltogether/v4-timelocks/blob/pool-1910-v4-timelocks-develop-new-timelock-smart/contracts/DrawAndPrizeDistributionTimelock.sol#L8

When PR is merged update `v4-timelocks`

**Comment by asselstine** [comment]
Yes!  Reviews would be great

**Review by PierrickGT — CHANGES_REQUESTED** [review]
Some minor changes and questions.
Also, we should remove the unused `newestDrawId` from the `PrizeTierHistory` contract.
https://github.com/pooltogether/v4-periphery/pull/9/files#diff-778c8645ac94af419acc3ce293c48834a903eb2376cf11e877a7db8591c6f5a9R56

File: test/PrizeDistributionFactory.test.ts [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            ).to.deep.include(
+                createPrizeDistribution({
+                    matchCardinality: 3,
+                    numberOfPicks: BigNumber.from(64),
+                }),
+            );
+        })
+    });
+
+    describe('pushPrizeDistribution()', () => {
```
Comment (PierrickGT) [inline test/PrizeDistributionFactory.test.ts:264]: Test that this function is only callable by the owner or manager.

File: test/PrizeDistributionFactory.test.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        it('should push the prize distribution onto the buffer', async () => {
+            await setupMocks();
+            await prizeDistributionBuffer.mock.pushPrizeDistribution.returns(true);
+            await expect(prizeDistributionFactory.pushPrizeDistribution(1, toWei('1000')))
+                .to.emit(prizeDistributionFactory, 'PushedPrizeDistribution')
+                .withArgs(1, toWei('1000'));
+        });
+    });
+
+    describe('setPrizeDistribution()', () => {
```
Comment (PierrickGT) [inline test/PrizeDistributionFactory.test.ts:274]: Test that this function is only callable by the owner and not the manager.

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: contracts/PrizeDistributionFactory.sol [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        require(address(_drawBuffer) != address(0), "PDC/db-zero");
+        require(address(_prizeDistributionBuffer) != address(0), "PDC/pdb-zero");
+        require(address(_ticket) != address(0), "PDC/ticket-zero");
+        require(_minPickCost > 0, "PDC/pick-cost-gt-zero");
+        
+        minPickCost = _minPickCost;        
+        prizeTierHistory = _prizeTierHistory;
+        drawBuffer = _drawBuffer;
+        prizeDistributionBuffer = _prizeDistributionBuffer;
+        ticket = _ticket;
```
[the remark this answers is not in the document]
Comment (asselstine) [reply contracts/PrizeDistributionFactory.sol:64]: I think it's superfluous for a standard contract with a constructor

File: contracts/PrizeDistributionFactory.sol [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        returns (IPrizeDistributionBuffer.PrizeDistribution memory)
+    {
+        IPrizeDistributionBuffer.PrizeDistribution
+            memory prizeDistribution = calculatePrizeDistribution(
+                _drawId,
+                _totalNetworkTicketSupply
+            );
+        prizeDistributionBuffer.pushPrizeDistribution(_drawId, prizeDistribution);
+
+        emit PushedPrizeDistribution(_drawId, _totalNetworkTicketSupply);
```
[the remark this answers is not in the document]
Comment (asselstine) [reply contracts/PrizeDistributionFactory.sol:86]: Yes, will rename if you like

If people are only listening to events then it's handy to have

File: contracts/PrizeDistributionFactory.sol [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        returns (IPrizeDistributionBuffer.PrizeDistribution memory)
+    {
+        IPrizeDistributionBuffer.PrizeDistribution
+            memory prizeDistribution = calculatePrizeDistribution(
+                _drawId,
+                _totalNetworkTicketSupply
+            );
+        prizeDistributionBuffer.setPrizeDistribution(_drawId, prizeDistribution);
+
+        emit SetPrizeDistribution(_drawId, _totalNetworkTicketSupply);
```
[the remark this answers is not in the document]
Comment (asselstine) [reply contracts/PrizeDistributionFactory.sol:110]: same same

File: contracts/PrizeDistributionFactory.sol [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        virtual
+        returns (IPrizeDistributionBuffer.PrizeDistribution memory)
+    {
+        IDrawBeacon.Draw memory draw = drawBuffer.getDraw(_drawId);
+        return calculatePrizeDistributionWithDrawData(_drawId, _totalNetworkTicketSupply, draw.beaconPeriodSeconds, draw.timestamp);
+    }
+
+    /**
+     * @notice Calculates what the prize distribution will be, given a draw id and total network ticket supply.
+     * @param _drawId The draw from which to use the Draw and
```
[the remark this answers is not in the document]
Comment (asselstine) [reply contracts/PrizeDistributionFactory.sol:132]: :+1:

**Pull request merged**